### PR TITLE
[DEV-13007] Integrate ModernBert 

### PR DESCRIPTION
**Pull request by benleetownsend**
TODOs

- [x] Verify features on a single short sample
- [x] Verify features on longer inputs to test local attention masking.
- [x] Verify features on batches of inputs to test padding.
- [x] Migrate the HF config items to finetune config.
- [x] Port ModernBert-Large
- [x] Push model files to S3
- [x] Move verification into a unit-test
- [x] Re-integrate gradient checkpointing
- [x] Find decent default hyperparams

### CODE REVIEW

**Comment by benleetownsend** [comment]
Need to finetune the predict batch sizes a bit for modern-bert but I think in general this is encouraging @madisonmay 

| dataset       | setting        | model        | training_time | evaluation_time | macro_f1 | macro_precision | macro_recall |
|--------------|---------------|--------------|---------------|----------------|----------|----------------|--------------|
| charities    | predict_speed | modern_bert  | 13628         | 543            | 0.737    | 0.741          | 0.752        |
| charities    | predict_speed | roberta      | 16891         | 213            | 0.695    | 0.650          | 0.775        |
| charities    | accuracy      | modern_bert  | 49469         | 307            | 0.736    | 0.757          | 0.736        |
| charities    | accuracy      | roberta      | 65767         | 700            | 0.698    | 0.644          | 0.810        |
| fcc_invoices | predict_speed | modern_bert  | 5932          | 412            | 0.777    | 0.675          | 0.934        |
| fcc_invoices | predict_speed | roberta      | 14446         | 414            | 0.761    | 0.655          | 0.929        |
| fcc_invoices | accuracy      | modern_bert  | 30301         | 707            | 0.775    | 0.671          | 0.947        |
| fcc_invoices | accuracy      | roberta      | 30744         | 483            | 0.763    | 0.656          | 0.935        |
| nda          | predict_speed | modern_bert  | 4329          | 56             | 0.872    | 0.841          | 0.907        |
| nda          | predict_speed | roberta      | 2086          | 43             | 0.849    | 0.798          | 0.908        |
| nda          | accuracy      | modern_bert  | 14835         | 91             | 0.881    | 0.870          | 0.897        |
| nda          | accuracy      | roberta      | 10758         | 80             | 0.856    | 0.824          | 0.892        |

File: docker-compose.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ services:
       dockerfile: docker/Dockerfile.gpu
     volumes:
       - .:/Finetune
+      - ../project_fruitfly/datasets:/datasets
```
Comment (benleetownsend) [inline docker-compose.yml]: remove

File: finetune/base_models/modern_bert/model.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        else:
+            raise ValueError(
+                "Cannot optimise hyperparams for {}, must be either 'speed', 'predict_speed' or 'accuracy'".format(
+                    config.optimize_for
+                )
+            )
+        return overrides
+
+
+class ModernBertLargeModel(SourceModel):
```
Comment (benleetownsend) [inline finetune/base_models/modern_bert/model.py:120]: Got much less love than ModernBert, it's just a bit too big to be practical for our uses. I think we throw this on the backlog and come back to it.

File: finetune/util/download.py [hunk 3]
```diff
@@ -13,6 +13,8 @@
 ROBERTA_BASE_URL = "https://s3.amazonaws.com/bendropbox/roberta/"
 OSCAR_BASE_URL   = "https://s3.amazonaws.com/bendropbox/oscar/"
 LAYOUTLM_BASE_URL   = "https://s3.amazonaws.com/bendropbox/layoutlm/"
+MODERN_BERT_BASE_URL   = "https://s3.amazonaws.com/bendropbox/modern_bert/"
```
Comment (madisonmay) [inline finetune/util/download.py:16]: Lol I forgot we used this

File: finetune/config.py [hunk 4]
```diff
@@ -348,6 +348,14 @@ def get_default_config():
         fusion_low_memory=True,
         include_bos_eos=True,
         #
+        # ModernBERT
+        norm_eps=1e-5,
```
Comment (madisonmay) [inline finetune/config.py:352]: I had forgotten about our mega-config

File: finetune/base_models/modern_bert/model.py (hunk 2 above)
Comment (madisonmay) [reply finetune/base_models/modern_bert/model.py:120]: Fine by me!

File: finetune/base_models/modern_bert/model.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        base_learning_rate = config.base_model.settings["lr"]
+        if config.optimize_for.lower() in ["accuracy", "accuracy_fp16"]:
+            overrides = {
+                "max_length": 2048,
+                "n_epochs": base_n_epochs,
+                "batch_size": 8,
+                "chunk_context": None,
+                "predict_batch_size": 16,
+                "mixed_precision": True,
+                "float_16_predict": True,
```
Comment (madisonmay) [inline finetune/base_models/modern_bert/model.py:96]: Basically all of our customers should be on compatible GPUs now?

**Review by madisonmay — APPROVED** [review]
